libcudacxx/include/cuda/ptx

```diff
@@ -0,0 +1,352 @@
+inline __device__ _CUDA_VSTD::uint64_t __as_gmem_ptr(const void * __ptr)   { return static_cast<_CUDA_VSTD::uint64_t>(__cvta_generic_to_global(__ptr)); }
+
+
+
+// SM 90 features
+// --------------
+
+/*
+ *  TMA / cp.async.bulk
+ *
+ */
+
+// cp.async.bulk
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-bulk
+
+// cp.reduce.async.bulk
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-reduce-async-bulk
+
+// cp.async.bulk.tensor
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-bulk-tensor
+
+// cp.reduce.async.bulk.tensor
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-reduce-async-bulk-tensor
+
+// cp.async.bulk.commit_group
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-bulk-commit-group
+
+// cp.async.bulk.wait_group
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-bulk-wait-group
+
+
+// Lower priority:
+
+// prefetch{.tensormap_space}.tensormap [a];  // prefetch the tensormap
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-prefetch-prefetchu
+
+// cp.async.bulk.prefetch
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-bulk
+
+// cp.async.bulk.prefetch.tensor
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cp-async-bulk-prefetch-tensor
+
+/*
+ *  Shared memory barrier
+ *
+ */
+
+// mbarrier.expect_tx
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-expect-tx
+
+// mbarrier.complete_tx
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-complete-tx
+
+// mbarrier.arrive.expect_tx
+// Support for count argument without the modifier .noComplete requires sm_90 or higher.
+// Qualifier .expect_tx requires sm_90 or higher.
+// Sub-qualifier ::cluster requires sm_90 or higher.
+// Support for .cluster scope requires sm_90 or higher.
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-arrive
+
+
+
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-arrive
+#if (defined(__CUDA_MINIMUM_ARCH__) && 900 <= __CUDA_MINIMUM_ARCH__) || (!defined(__CUDA_MINIMUM_ARCH__))
```

Comment:
I would prefer if we would immediately split each new architecture into its own file.

That will hopefully keep things on a somewhat sane level with increasing number of architectures

Reply:
That was my initial idea as well. Here is my thought process up to this point. I would be interested to hear your opinion. 

CUTLASS uses `<feature>_<architecture>.h` as a header file template. This gives a good overview of what is new in each architecture, but has the risk of dispersing related instructions across different headers. For instance, `mbarrier.test_wait` was introduced in sm80, and `mbarrier.try_wait` was introduced in sm90. It would be nice if they could be together in the same header, let's say `mbarrier.h`. 

I was also considering a user who just wants the latest PTX exposure for a quick experiment (either a benchmark, or in compiler explorer). They might want to just copy and paste the latest PTX header into their editor and go from there. They would benefit if the whole PTX exposure was just a single file. I hope this is achievable in practice, even though we use a lot of libcu++ specific macros and include paths.

So, I think the following two options are the most appealing:

1. Split into separate headers by feature (`mbarrier`) and / or by architecture (for grab bag features like `elect.sync`, `griddepcontrol`). This would benefit discoverability and maintainability of the code files.
2. Put everything in one big header. This would benefit users who want to copy-and-paste into compiler explorer and would put a bigger burden on maintenance.

Reply:
I would also dread fragmentation from a feature *AND* architecture split.

My gut feeling would be that it is easiest to separate by `architecture` and also make it so that each architecture header also includes all precluding headers, so that a user that "just" includes SM_90 will get all the instructions that are available up to SM_90

I am not sure, whether we actually want to deprecate PTX instructions, but this is something that could be done easily via a conditional [[deprecated]] annotation. SO I do not see any problem there.

Does that sound good for you?

Reply:
My idea would be to have only `<cuda/ptx>` as a public header. What is the use case you had in mind for including a header with PTX exposure up to a specific architecture? 

One thing to keep in mind is that PTX gets additions in minor releases as well. For instance, PTX 7.5 added "Support for ignore-src operand on cp.async instruction."  Therefore, if your goal is to minimize the number of occurrences of 
```
#if (defined(__CUDA_MINIMUM_ARCH__) && 900 <= __CUDA_MINIMUM_ARCH__) || (!defined(__CUDA_MINIMUM_ARCH__))
```
then I think this will be difficult, and in practice it may even be worse as there will also be some checks for specific CUDA versions.

> I am not sure, whether we actually want to deprecate PTX instructions

I am missing context here. Was this intended as a reply to my comment?

Reply:
My fear is that we end up with one gargantuan file. 

I am totally fine with us fixing this problem when it actually arises and not start with premature optimization from my side

Reply:
I agree with Michael, I'd like to see some form of organization, but I understand doing so based on architecture will likely end up having surprisingly high maintenance burden. 

If we did lexicographical organization of 'family, function, architecture' maybe it would be doable.  I'd like to talk about this sometime if we decide to do cleanup.

Reply:
The public facing header can just be a single `#include <cuda/ptx>` but the implementation of that header could be:
```
#include <detail/ptx/sm70>
#include <detail/ptx/sm75>
#include <detail/ptx/sm80>
...
```
So the implementation could still be broken up and better organized.

Reply:
I agree that the code organization was not optimal and have changed my opinion based on your feedback and discussions with others.

- I agree that the header should be broken up.
  - I think that trying to accomodate copying and pasting the header into compiler explorer is going to be too difficult.
- Inventing new taxonomies based on architecture or features is going to be a pain.
- Therefore, I took over the organization that was already there in the PTX ISA docs, and pasted the headers and links into the current PTX header, even for instructions that are not implemented. 
  - This makes it very clear where any PTX binding should be implemented
  - Certain sets of instructions, such as mbarrier may be separated into their own header to keep the code organized.